plugins/sources/interval_timer_test.go

```diff
@@ -0,0 +1,75 @@
+package sources
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestWaitInterval(t *testing.T) {
+	t.Run("zero time", func(t *testing.T) {
+		timer := NewIntervalTimer(15 * time.Second)
+		assert.Equal(t, 15*time.Second, timer.waitToNextInterval(0))
+	})
+
+	t.Run("inside first interval", func(t *testing.T) {
+		timer := NewIntervalTimer(15 * time.Second)
+		assert.Equal(t, 10*time.Second, timer.waitToNextInterval(5*time.Second))
+	})
+
+	t.Run("outside first interval", func(t *testing.T) {
+		timer := NewIntervalTimer(15 * time.Second)
+		assert.Equal(t, 14*time.Second, timer.waitToNextInterval(16*time.Second))
+	})
+
+	t.Run("many intervals late", func(t *testing.T) {
+		timer := NewIntervalTimer(15 * time.Second)
+		assert.Equal(t, 14*time.Second, timer.waitToNextInterval(61*time.Second))
+	})
+
+	t.Run("near but still before the interval", func(t *testing.T) {
+		timer := NewIntervalTimer(15 * time.Second)
+		assert.Equal(t, 15499*time.Millisecond, timer.waitToNextInterval(14501*time.Millisecond))
+	})
+}
+
+func TestReset(t *testing.T) {
+	t.Run("lastResetTime is set", func(t *testing.T) {
+		timer := NewIntervalTimer(15 * time.Second)
+		timer.Reset()
+		assert.NotZero(t, timer.lastResetTime)
+	})
+}
+
+func TestMissedCount(t *testing.T) {
+	t.Run("zero time", func(t *testing.T) {
+		timer := NewIntervalTimer(15 * time.Second)
+		assert.Equal(t, int64(0), timer.intervalsMissed())
+	})
+
+	t.Run("first interval", func(t *testing.T) {
+		timer := NewIntervalTimer(15 * time.Second)
+		timer.lastResetTime = time.Now().Add(-15 * time.Second)
```

Comment:
It seems that there is an implicit dependency that these tests work because the intervals in the tests are big enough to cover any discrepancies in the 3 calls to time.Now(). Might be worth a comment.